### ISSUE TITLE
Unable to install Nvidia GPU Operator in Openshift

### ISSUE THREAD

**Issue by altruistcoder** [NVIDIA/gpu-operator#157]
Hello, 

I am trying to install the Nvidia GPU Operator on my Openshift cluster but when I click on install, it keeps waiting for a long period of time and after a certain period of time, it gets failed.

When I tried searching for any errors that are occurring, I found the error shown in the below screenshot in the events section of the corresponding GPU Pod:

![image](https://user-images.githubusercontent.com/31319270/109625136-f60e2400-7b64-11eb-9997-0fad7436241f.png)

Also, I see this when I visit the GPU Operator details page:

![image](https://user-images.githubusercontent.com/31319270/109625229-14741f80-7b65-11eb-942e-b787e3d048ef.png)

**Openshift Cluster Version: 4.6.8
Nvidia GPU Operator Version: 1.6.0**

The GPU operator was working fine in our cluster till today morning when its version was 1.5.* already, but when we uninstalled the operator today and reinstalled it today again, we started facing this issue due to which we are not able to use GPUs on our cluster anymore.

Can you please help in resolving this issue?
@klueska @shivamerla

**Comment by kpouget** [comment]
Hello @altruistcoder,

the reference to `nvidia-container-cli.real` in unexpected for the Operator container, maybe something went wrong when you uninstalled the previous version?

could you try to run these commands:
```
oc describe pod/gpu-operator-55987fc888-mbzqb -n openshift-operators
oc logs  pod/gpu-operator-55987fc888-mbzqb -n openshift-operators # shouldn't work
```
```
oc get ClusterServiceVersion -A | grep gpu-operator
```
```
oc get pods -n gpu-operator-resources
```

**Comment by elezar** [comment]
Hi @altruistcoder.  Do you have any other details as to your configuration. There is a possible issue with containerd with empty configuration files under 1.6.0 that we are working on resolving. (Thanks for jumping in @kpouget )

**Comment by klueska** [comment]
Try removing any references to the nvidia-container-runtime in your /etc/docker/daemon.json file. That error is coming from the fact that your system still thinks it should be using the nvidia runtime instead of the normal runc. That should be cleaned up when uninstalling the operator, but it looks like maybe something went wrong.

**Comment by kpouget** [comment]
@elezar,
> There is a possible issue with containerd

OCP shouldn't be using containerd, only crio, right?

FYI, we tested successfully 1.6.0 on OCP 4.6 last night, so there shouldn't be any major issue with this release, that's why I'm thinking the error could come from the uninstall step

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-psap-ci-artifacts-release-4.6-gpu-operator-e2e/1366523719807143936/artifacts/gpu-operator-e2e/nightly/artifacts/gpu-ci/extra_logs/gpu_pods_list
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-psap-ci-artifacts-release-4.6-gpu-operator-e2e/1366523719807143936/artifacts/gpu-operator-e2e/nightly/build-log.txt

**Comment by klueska** [comment]
Right... then I amend my comment to say you should delete the hook from the crio hooks directory...

**Comment by altruistcoder** [comment]
> Hello @altruistcoder,
> 
> the reference to `nvidia-container-cli.real` in unexpected for the Operator container, maybe something went wrong when you uninstalled the previous version?
> 
> could you try to run these commands:
> 
> ```
> oc describe pod/gpu-operator-55987fc888-mbzqb -n openshift-operators
> oc logs  pod/gpu-operator-55987fc888-mbzqb -n openshift-operators # shouldn't work
> ```

Hii @kpouget , I have attached a file which contains the output of the "oc describe command" for the respective GPU pod:

[gpu-operator-55987fc888-mbzqb-description.txt](https://github.com/NVIDIA/gpu-operator/files/6067662/gpu-operator-55987fc888-mbzqb-description.txt)

**Comment by altruistcoder** [comment]
> Hello @altruistcoder,
> 
> the reference to `nvidia-container-cli.real` in unexpected for the Operator container, maybe something went wrong when you uninstalled the previous version?
> 
> could you try to run these commands:
> 
> ```
> oc get ClusterServiceVersion -A | grep gpu-operator
> ```

I also ran this command to get the list of Cluster Service Versions and I saw that in almost all of them the "PHASE" is coming as "Failed".

**Comment by kpouget** [comment]
thanks @altruistcoder for the extra info, it seems to be clear now that the issue is linked with the uninstall of the previous [version](MachineConfig) 

@klueska my `` looks like this:
```
# oc debug node/$NODE -t -- chroot /host cat /etc/docker/daemon.json
{
  "runtimes": {
    "nvidia": {
      "path": "/usr/local/nvidia/toolkit/nvidia-container-runtime"
    }
  },
  "default-runtime": "nvidia"
}
```
should @altruistcoder edit it to ```{}``` ?

**Comment by klueska** [comment]
Well if `cri-o` is used and not `docker`, then he should actually be removing the following file instead of editing `/etc/docker/daemon.json`.
```
/usr/share/containers/oci/hooks.d/oci-nvidia-hook.json
```

**Comment by altruistcoder** [comment]
Thanks a lot for responding  @klueska and @kpouget 

I tried searching for the files specified by @klueska in my master node but I am not able to see any files under `/usr/share/containers/oci/hooks.d/` directory and I also didn't get this file: `/etc/docker/daemon.json`

**Comment by kpouget** [comment]
@altruistcoder the correct path (on the host) is `/etc/containers/oci/hooks.d/oci-nvidia-hook.json` which contains in my cluster:

```
{
    "version": "1.0.0",
    "hook": {
        "path": "/usr/local/nvidia/toolkit/nvidia-container-toolkit",                           
        "args": ["nvidia-container-toolkit", "prestart"],                                       
        "env": [
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/nvidia/toolkit"
        ]
    },
    "when": {
        "always": true,
        "commands": [".*"]
    },
    "stages": ["prestart"]
}
```

> Well if cri-o is used and not docker, then he should actually be removing the following file instead of editing /etc/docker/daemon.json.

hum, maybe then `/etc/docker/daemon.json` shouldn't populated? (or maybe it's a left-over from an older operator version in my system)

**Comment by kpouget** [comment]
@altruistcoder 
> I tried searching for the files specified by klueska in my master node 
it's not on the master node but on the GPU nodes

**Comment by altruistcoder** [comment]
It worked!!

Thanks a lot @klueska @kpouget and @elezar for helping me in resolving this issue.

**Comment by kpouget** [comment]
> Can you please suggest me on what I might have done wrong during the uninstallation of the Operator and what is the right method to do it?

I'm not sure you did anything wrong, but at the same time I'm unable to reproduce it with the GPU Operator 1.6.0 on OCP 4.7,
and I actually uninstalled the GPU Operator many many times in the past, so I wonder what may have happened in your cluster!

**Comment by shivamerla** [comment]
@altruistcoder Did you by any chance force reboot the node before previous un-install. I can imagine this happening with force reboot of node, where container-toolkit will not be able to remove the runtime hook. During graceful shutdown/reboot, it will cleanup this properly. When this happens on reboot none of the operator pods can come up. We are working on solution for this.

**Comment by altruistcoder** [comment]
Actually, we stop our cluster in the night and start it in the morning using a scheduled lambda function (using a boto3 script) which does this job daily. 
I think this way of rebooting the cluster might be taken as the Graceful Reboot of the cluster as can be inferred from the information presented [here](https://docs.openshift.com/container-platform/4.6/backup_and_restore/graceful-cluster-shutdown.html) as well.
But if it is not, I would be really happy if you could correct me and suggest me a better way to do it so that I don't have to face such problems in the future.

**Comment by altruistcoder** [comment]
Hello @klueska @kpouget,

I am still facing the same problem which I reported initially in this issue. The only difference it is not happening now in all my nodes but in some of them randomly. As I stated in my previous message, our cluster is scheduled to stop every night and start in the morning. So, when the cluster gets started in the morning, the Nvidia GPU Pods (viz, Driver, DevicePlugin, DCGM, Container-toolkit, GPU-Feature-Discovery Pods) of some of the nodes are in error state and I have to again perform the steps that you suggested that day which is deleting the file located at `/etc/containers/oci/hooks.d/oci-nvidia-hook.json` from the corresponding node.

Will I have to do this task everyday when needed from now onwards or is there any permanent fix for this?

**Comment by kpouget** [comment]
I wanted to reproduce the situation by 
1/ uninstalling the GPU Operator
2/ creating `/etc/containers/oci/hooks.d/oci-nvidia-hook.json` with the content from  https://github.com/NVIDIA/gpu-operator/issues/157#issuecomment-788801531
3/ reinstall the operator

I was expecting the creation of any Pod on this node to fail (because the nvidia-runtime isn't setup), but to my surprise, the operator & daemonset deployed without any issue.

@klueska do you think I missed a step?

**Comment by shivamerla** [comment]
@altruistcoder We will be fixing it soon with next update. Meanwhile you can edit the toolkit pod with below change in the initContainer. It will make sure to cleanup the nvidia CRI hook before drivers are loaded.

```
diff --git a/assets/state-container-toolkit/0400_container_toolkit.yml b/assets/state-container-toolkit/0400_container_toolkit.yml
index 1247a92f..829caf65 100644
--- a/assets/state-container-toolkit/0400_container_toolkit.yml
+++ b/assets/state-container-toolkit/0400_container_toolkit.yml
@@ -32,7 +32,7 @@ spec:
         export NVIDIA_LIBRARY_PATH=/run/nvidia/driver/usr/lib/x86_64-linux-gnu/:/run/nvidia/driver/usr/lib64; \
         export LD_LIBRARY_PATH=${SYS_LIBRARY_PATH}:${NVIDIA_LIBRARY_PATH}; echo ${LD_LIBRARY_PATH}; \
         export PATH=/run/nvidia/driver/usr/bin/:${PATH}; \
-        until nvidia-smi; do echo waiting for nvidia drivers to be loaded; sleep 5; done"]
+        until nvidia-smi; do echo waiting for nvidia drivers to be loaded; rm -f /usr/share/containers/oci/hooks.d/oci-nvidia-hook.json; sleep 5; done"]
         securityContext:
           privileged: true
           seLinuxOptions:
@@ -41,6 +41,8 @@ spec:
           - name: nvidia-install-path
             mountPath: /run/nvidia
             mountPropagation: Bidirectional
+          - name: crio-hooks
+            mountPath: /usr/share/containers/oci/hooks.d
```

**Comment by altruistcoder** [comment]
@shivamerla Pardon me but I am not able to understand where exactly I need to make these changes as I am installing the Nvidia GPU Operator from Openshift's OperatorHub.
Also, if I am able to make the above changes successfully at the right location, will these changes solve the above main error and the pods would work properly even if I ever need to reinstall the Operator again due to any reason or in my scenario specifically where my cluster is stopped & started again on a daily basis?

**Comment by shivamerla** [comment]
@altruistcoder You can edit nvidia-container-toolkit daemonset spec under `gpu-operator-resources` project from the UI.

![Screen Shot 2021-03-05 at 7 22 19 AM](https://user-images.githubusercontent.com/30056107/110135765-bf622480-7d83-11eb-9bb6-588a3acd03a8.png)

After editing, click save and changes will be applied. You can ignore the warning and check that pods are started again. This change will handle the reboot issue you are seeing. We will be fixing this with next bundle update, so on update you will not need to apply this workaround.

**Comment by zvonkok** [comment]
@shivamerla @klueska Going forward we should place the hook to `/run/containers/oci/hooks.d` this will prevent the hook from being persistent e.g. after a reboot.

**Comment by shivamerla** [comment]
@zvonkok good point. Will check with @klueska if we can make that change.

**Comment by shivamerla** [comment]
PR for this: https://gitlab.com/nvidia/kubernetes/gpu-operator/-/merge_requests/192

**Comment by shivamerla** [comment]
@altruistcoder  This is fixed in 1.6.2 published to catalog. Please un-install old one and deploy this. This will place nvidia runtime hook configuration under `/run/containers/oci/hooks.d` and  solves reboot issue you were facing.

**Comment by altruistcoder** [comment]
Thanks a lot @shivamerla for providing a solution to this problem so quickly. I have updated my Nvidia GPU Operator to version 1.6.2 in my cluster and I believe I'll not face this problem anymore.

Thank You @kpouget @klueska for helping in resolving this error.
